Sources/SwiftDocCUtilities/Action/Actions/MergeAction.swift

```diff
@@ -0,0 +1,65 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2024 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+@_spi(FileManagerProtocol) import SwiftDocC
+
+/// An action that merges a list of documentation archives into a combined archive.
+struct MergeAction: Action {
+    var archives: [URL]
+    var landingPageCatalog: URL?
+    var outputURL: URL
+    var fileManager: FileManagerProtocol
+
+    mutating func perform(logHandle: LogHandle) throws -> ActionResult {
+        guard let firstArchive = archives.first else {
+            // A validation warning should have already been raised in `Docc/Merge/InputAndOutputOptions/validate()`.
+            return ActionResult(didEncounterError: true, outputs: [])
+        }
+        
+        try? fileManager.removeItem(at: outputURL)
+        try fileManager.copyItem(at: firstArchive, to: outputURL)
+        
+        // TODO: Merge the LMDB navigator index
+        
+        let jsonIndexURL = outputURL.appendingPathComponent("index/index.json")
+        guard let jsonIndexData = fileManager.contents(atPath: jsonIndexURL.path) else {
+            // TODO: Error
+            return ActionResult(didEncounterError: true, outputs: [])
+        }
+        var combinedJSONIndex = try JSONDecoder().decode(RenderIndex.self, from: jsonIndexData)
+        
+        for archive in archives.dropFirst() {
+            for directoryToCopy in ["data/documentation", "data/tutorials", "documentation", "tutorials", "images", "videos", "downloads"] {
+                let fromDirectory = archive.appendingPathComponent(directoryToCopy, isDirectory: true)
+                let toDirectory = outputURL.appendingPathComponent(directoryToCopy, isDirectory: true)
+                
+                for from in (try? fileManager.contentsOfDirectory(at: fromDirectory, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)) ?? [] {
+                    try fileManager.copyItem(at: from, to: toDirectory.appendingPathComponent(from.lastPathComponent))
+                }
+            }
+            guard let jsonIndexData = fileManager.contents(atPath: archive.appendingPathComponent("index/index.json").path) else {
+                // TODO: Error
+                return ActionResult(didEncounterError: true, outputs: [])
+            }
+            let renderIndex = try JSONDecoder().decode(RenderIndex.self, from: jsonIndexData)
+            
+            try combinedJSONIndex.merge(renderIndex)
```

Comment:
Can we be sure that none of the archives will have overlapping paths? Should we validate this before merging the indexes like this? And before copying around all of the files above?

Probably we should have a "pre-flight" check... meaning we iterate over all the files to be merged, to be sure none of them overlap or conflict. If they do, we can report an error before actually copying any of them.

Or alternatively, as mentioned in an earlier comment, if we use a temporary directory as a staging area, then we could abort the merge if we find any conflicting paths, before creating the output path and writing anything into it.

Reply:
Fixed in https://github.com/apple/swift-docc/pull/821/commits/f98be845a62d75c7d8d019587e09793abd0ad12a with a new `testErrorWhenArchivesContainOverlappingData` test